File: src/wholist.c
/*-------------------------------------------------------------------------
cvsinfo: $Header$
Maarten's Mud, WWW-based MUD using MYSQL
Copyright (C) 1998  Maarten van Leunen

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

Maarten van Leunen
Driek van Erpstraat 9
5341 AK Oss
Nederland
Europe
maartenl@il.fontys.nl
-------------------------------------------------------------------------*/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <time.h>
#include <stdarg.h>
#include <sys/file.h>
#include "mysql.h"

#define debuggin 0

void exiterr(int exitcode, char *sqlstring, MYSQL *mysql)
{
printf("Error %i: %s\n {%s}\n", exitcode, mysql_error(mysql), sqlstring );
}

int SendSQL()
{
	MYSQL mysql;
	MYSQL_RES *res;
	MYSQL_ROW row;
	FILE *fp;
	uint i = 0;
	
	char sqlstring[1024];
	strcpy(sqlstring, "select name, title, "
	"time_to_sec(date_sub(NOW(), INTERVAL 2 HOUR))-time_to_sec(lastlogin)"
	", sleep from tmp_usertable where god<=1");
	 
	if (!(mysql_connect(&mysql,"localhost","root", ""))) 
		exiterr(1, sqlstring, &mysql);
 
	if (mysql_select_db(&mysql,"mud"))
		exiterr(2, sqlstring, &mysql);
 
	if (mysql_query(&mysql,sqlstring))
		exiterr(3, sqlstring, &mysql);
 
	if (!(res = mysql_store_result(&mysql)))
	{
		exiterr(4, sqlstring, &mysql);
	} 
	else 
	{
		printf("Content-type: text/html\n\n");
		printf("<META http-equiv=\"refresh\" content=\"120\">");
		printf("<META HTTP-EQUIV=\"pragma\" CONTENT=\"no-cache\">\r\n\r\n");
		printf("<HTML>\n");
		printf("<HEAD>\n");
		printf("<TITLE>\n");
		printf("Land of Karchan - Who\n");
		printf("</TITLE>\n");
		printf("</HEAD>\n");
		
		printf("<BODY>\n");
		printf("<BODY BGCOLOR=#FFFFFF BACKGROUND=\"/images/gif/webpic/back4.gif\" >\n");
		printf("<H2>");
		printf("<IMG SRC=\"http://www.karchan.org/images/gif/dragon.gif\">");
		printf("List of All Active Users</H2>\r\n");
		printf("<I>There are %i persons active in the game.</I><P>\r\n",
			mysql_num_rows(res));
		printf("<UL>");
		while((row = mysql_fetch_row(res))) {
			if (atoi(row[3])==1)
			{
				printf("<LI>%s, %s, sleeping (%i min, %i sec idle)\r\n", 
				row[0], row[1], atoi(row[2]) / 60, atoi(row[2]) % 60);
			}
			else
			{
				printf("<LI>%s, %s (%i min, %i sec idle)\r\n", 
				row[0], row[1], atoi(row[2]) / 60, atoi(row[2]) & 60);
			}
		}
		printf("</UL>\r\n");

		printf("</BODY>\r\n");
		printf("</HTML>\r\n");
	}
	if (!mysql_eof(res))
		exiterr(5, sqlstring, &mysql);
 
	mysql_free_result(res);
	mysql_close(&mysql);
}

int
main()
{
	SendSQL();
	return 0;
}

File: scripts/wholist.php
<?
/*-------------------------------------------------------------------------
svninfo: $Id$
Maarten's Mud, WWW-based MUD using MYSQL
Copyright (C) 1998  Maarten van Leunen

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

Maarten van Leunen
Appelhof 27
5345 KA Oss
Nederland
Europe
maarten_l@yahoo.com
-------------------------------------------------------------------------*/
?>
<META http-equiv="refresh" content="120">
<META HTTP-EQUIV="pragma" CONTENT="no-cache">
<HTML>
<HEAD>
<TITLE>
Land of Karchan - Who
</TITLE>
</HEAD>
                                                                                                                      
<BODY>
<BODY BGCOLOR=#FFFFFF BACKGROUND="/images/gif/webpic/back4.gif">
<H2>
<IMG SRC="/images/gif/dragon.gif">
List of All Active Users</H2>
<?php
include $_SERVER['DOCUMENT_ROOT']."/scripts/connect.php"; 
$result = mysql_query("select name, title, sleep, 
	floor((unix_timestamp(NOW())-unix_timestamp(lastlogin)) / 60) as min,
	((unix_timestamp(NOW())-unix_timestamp(lastlogin)) % 60) as sec,
	if (mm_area.area <> \"Main\", concat(\" in \" , mm_area.shortdesc), \"\") as area
	from mm_usertable, mm_rooms, mm_area 
	where god<=1 and active=1 and mm_rooms.id = mm_usertable.room and
	mm_rooms.area = mm_area.area"
	, $dbhandle)
	or error_message("Query failed : " . mysql_error());
//	error_message("Query failed : " . mysql_error());
printf("<I>There are %s persons active in the game.</I><P><UL>\r\n", 
	mysql_num_rows($result));
while ($myrow = mysql_fetch_array($result)) 
{
	if ($myrow[2] == "1")
	{
		printf("<li>%s, %s, sleeping %s (logged on %s min, %s sec ago)\r\n",
		$myrow["name"], $myrow["title"], $myrow["area"], $myrow["min"], $myrow["sec"]);
	}
	else
	{
		printf("<li>%s, %s %s (logged on %s min, %s sec ago)\r\n",
		$myrow["name"], $myrow["title"], $myrow["area"], $myrow["min"], $myrow["sec"]);
	}

}
mysql_close($dbhandle);
?>
</UL>
<hr>

<script language="JavaScript">

<!-- In hiding!
 browserName = navigator.appName;          
           browserVer = parseInt(navigator.appVersion);
               backon = new Image;          
               backon.src = "/images/gif/webpic/new/buttono.gif";
               
               
               backoff = new Image;
               backoff.src = "/images/gif/webpic/buttono.gif";
               

function img_act(imgName) {
        imgOn = eval(imgName + "on.src");
        document [imgName].src = imgOn;
}

function img_inact(imgName) {
        imgOff = eval(imgName + "off.src");
        document [imgName].src = imgOff;
}

//-->
</SCRIPT>
<P><BR><BR>
<A HREF="/karchan/index.html" onMouseOver="img_act('back')" onMouseOut="img_inact('back')">
<IMG ALT="Backitup!" SRC="/images/gif/webpic/buttono.gif" BORDER="0" name="back"><br></A> </TD>
<DIV ALIGN=right>Last Updated $Date$
</DIV>
</BODY>
</HTML>
